Add approach_speed field to Location message

luca-della-vedova opened the pull request:
## New feature implementation

### Implemented feature

This PR allows simulated robots to respect speed limits. Speed limits for lanes are kept track of by upstream nodes (i.e. fleet adapters / managers) and published in the robot_path_requests message through a new field `approach_speed`. For each location `approach_speed` denotes the speed limit of the lane leading to it.

Yadunund commented on rmf_fleet_msgs/msg/Location.msg:
```diff
@@ -2,5 +2,6 @@ builtin_interfaces/Time t
 float32 x
 float32 y
 float32 yaw
+float32 approach_speed # Target speed of the lane leading to this waypoint
```
Maybe the comment can specify what range of values will enforce speed limits? Eg: greater than zero

aaronchongth replied:
@luca-della-vedova I missed this, it would be good to add a comment about limits and the units as well

luca-della-vedova replied:
The fact that it was a limit was stated after the feedback but I added the unit (m/s) https://github.com/open-rmf/rmf_internal_msgs/pull/28/commits/b50f98319619780e36af4dd4aa6c5e841b08cf19

aaronchongth commented on rmf_fleet_msgs/msg/Location.msg:
```diff
@@ -2,5 +2,9 @@ builtin_interfaces/Time t
 float32 yaw
+
+# Target speed of the lane leading to this waypoint, to be respected if > 0.0
+float32 approach_speed
```
Since the field is about speed limits, I would hesitate to rely on zero initialization. Could we have the default value to be -1 instead, and change the field name to be `approach_speed_limit` too?

The documentation makes it sound like the robots **must** be travelling at this speed approaching this location.

aaronchongth commented:
On second thought, after seeing how it's used in https://github.com/open-rmf/rmf_ros2/pull/132, would it be better to emulate an optional field with

```
bool obey_approach_speed_limit = false
float32 approach_speed_limit
```

This will prevent any ambiguity, floating point errors, and also allow us to check for errors easier down the line. Let me know what you think.

luca-della-vedova commented:
> On second thought, after seeing how it's used in [open-rmf/rmf_ros2#132](https://github.com/open-rmf/rmf_ros2/pull/132), would it be better to emulate an optional field with
> 
> ```
> bool obey_approach_speed_limit = false
> float32 approach_speed_limit
> ```
> 
> This will prevent any ambiguity, floating point errors, and also allow us to check for errors easier down the line. Let me know what you think.

Done https://github.com/open-rmf/rmf_internal_msgs/pull/28/commits/6d805f3412212440b420b18aedeb9b471e4cb0cd
It seems booleans are [initialized to false](https://design.ros2.org/articles/legacy_interface_definition.html) by default but I added it for extra clarity. Updated the [matching PR in rmf_simulation](https://github.com/open-rmf/rmf_simulation/pull/56) to account for the change as well, didn't do the `rmf_ros2` since it seems we might drop it and just have the new feature in the [new fleet adapter](https://github.com/open-rmf/rmf_demos/pull/109)

aaronchongth reviewed (approved):
This looks great! LGTM

aaronchongth reviewed (approved):
Thanks for the quick fix.
